### PR TITLE
Add code to insert transition record; code re-factoring

### LINKED ISSUE #11914: Adjust records in MSPileup DB for support partial pileup data placement
**Impact of the new feature**
Prepare MSPIleup database for handling partial pileup data placement

**Is your feature request related to a problem? Please describe.**
At the moment all existing documents in MSPileup DB contains empty transition record list. The new MSPileup code base requires to have full transition record at the time we need to update it. Current the code creates it on its own but we need to adjust existing records to align with it.

**Describe the solution you'd like**
Create a new script and run it on production cmsweb cluster within MSPileup MongoDB. It should have the following logic:
- Get list of existing pileup records in MongoDB
- Loop over records 
- For each record extract pileup name and 
- Create new transition record with the following content: {‘containerFraction’:1, ‘customDID’: <pileupname>, ‘updateTime’: gmtimeSeconds(), ‘userDN’: <userDN>}
- We must use WM robot DN for userDN value, e.g. `/DC=ch/DC=cern/OU=Organic Units/OU=Users/CN=cmsunified/CN=658085/CN=Robot: CMS Unified account`
- We must use `gmtimeSeconds()` function or properly determine number seconds since epoch in GMT time zone

**Describe alternatives you've considered**
None

**Additional context**

### PR DESCRIPTION
**Pull request by vkuznet**
Fixes #11914 

#### Status
ready

#### Description
Update ad-hoc scripts to create initial transition record for MSPileup records. There are two scripts:
- `adjustMongoDocs.py` to communicate with MongoDB directly
- `updatePileupObjects.py` to communicate with MongoDB backend via MSPileup REST API

Here is how each script can be called:
```
# adjust port, dbname and dbcoll appropriately
adjustMongoDocs.py --add-tran-record --dburi=localhost:8230 --dbname=test --dbcoll=test --verbose 1

# use update script to rely on MSPileup REST API
updatePileupObjects.py --add-tran-record -url=https://cmsweb.cern.ch --userDN=<user-dn>
```

#### Is it backward compatible (if not, which system it affects?)
YES

#### Related PRs
part of meta https://github.com/dmwm/WMCore/issues/11537

#### External dependencies / deployment changes
pymongo client

### CODE REVIEW

File: bin/adhoc-scripts/adjustMongoDocs.py [hunk 1]
```diff
@@ -25,12 +28,16 @@
...
+# WMCore services
+from Utils.Timers import gmtimeSeconds
+
 
 class OptionParser(object):
     "Helper class to handle script options"
 
     def __init__(self):
         "User based option parser"
+        defaultDN = '/DC=ch/DC=cern/OU=Organic Units/OU=Users/CN=cmsunified/CN=658085/CN=Robot: CMS Unified account'
```
Comment (amaltaro) [inline bin/adhoc-scripts/adjustMongoDocs.py:40]: I would suggest to use `/DC=ch/DC=cern/OU=Organic Units/OU=Users/CN=msunmer/CN=852819/CN=Robot: WmCore Service Account`.

**Review by amaltaro — CHANGES_REQUESTED** [review]
Instead of updating this script which requires direct access to the mongodb cluster, I would suggest to provide such changes under: https://github.com/dmwm/WMCore/blob/master/bin/adhoc-scripts/updatePileupObjects.py

File: bin/adhoc-scripts/adjustMongoDocs.py (hunk 1 above)
Comment (vkuznet) [reply bin/adhoc-scripts/adjustMongoDocs.py]: ok

File: bin/adhoc-scripts/updatePileupObjects.py [hunk 2]
```diff
@@ -33,6 +40,10 @@ def __init__(self):
         self.parser.add_argument("--url", action="store",
                                  dest="url", default="https://cmsweb-testbed.cern.ch",
                                  help="URL for the MSPileup service")
+        self.parser.add_argument("--userdn", action="store",
+                                 dest="userDN", default=defaultDN, help="user DN")
+        self.parser.add_argument("--add-tran-record", action="store_true",
+                                 dest="transition", help="add transaction record")
```
Comment (amaltaro) [inline bin/adhoc-scripts/updatePileupObjects.py:46]: typo in this line

File: bin/adhoc-scripts/updatePileupObjects.py [hunk 3]
```diff
@@ -179,23 +190,70 @@ def main():
...
+    :param url: url to use (string)
+    :param userDN: string representing user DN
+    :param logger: logger to use
+    :param dryrun: option to run dry-run mode (boolean)
+    :return: nothing
+    """
+    puDocs = getPileupDocs(url, handler, logger)
+    logger.info("Found %d documents in MSPileup", len(puDocs))
+    for rec in puDocs:
+        if 'pileupName' not in rec:
```
Comment (amaltaro) [inline bin/adhoc-scripts/updatePileupObjects.py:212]: is this case even possible? if not, then please just drop it.

File: bin/adhoc-scripts/updatePileupObjects.py [hunk 4]
```diff
@@ -179,23 +190,70 @@ def main():
...
+            continue
+        if ('transition' in rec and len(rec['transition']) == 0) or 'transition' not in rec:
```
Comment (amaltaro) [inline bin/adhoc-scripts/updatePileupObjects.py:214]: This line can be refactored to `if not rec.get("transition"):`.

File: bin/adhoc-scripts/updatePileupObjects.py [hunk 5]
```diff
@@ -179,23 +190,70 @@ def main():
...
+
+    if dryrun:
+        logger.info("documents are not written due to --dry-run option\n")
+    else:
+        # finally, update the pileup documents in the database
+        writePileupDocs(url, puDocs, handler, logger)
+        logger.info("documents are uploaded to MSPileup\n")
+
+
+def updatePileupRecords(handler, url, fin, logger, dryrun):
```
Comment (amaltaro) [inline bin/adhoc-scripts/updatePileupObjects.py:230]: I would suggest moving these 2 new methods above the `main` function.

**Review by amaltaro — CHANGES_REQUESTED** [review]

File: bin/adhoc-scripts/updatePileupObjects.py (hunk 2 above)
Comment (vkuznet) [reply bin/adhoc-scripts/updatePileupObjects.py]: fixed

File: bin/adhoc-scripts/updatePileupObjects.py (hunk 3 above)
Comment (vkuznet) [reply bin/adhoc-scripts/updatePileupObjects.py]: this is (was) useful for local testing, but in cmsweb and cmsweb-testbed of course it is not necessary.

File: bin/adhoc-scripts/updatePileupObjects.py (hunk 4 above)
Comment (vkuznet) [reply bin/adhoc-scripts/updatePileupObjects.py]: fine.

File: bin/adhoc-scripts/updatePileupObjects.py (hunk 5 above)
Comment (vkuznet) [reply bin/adhoc-scripts/updatePileupObjects.py:190]: moved

File: bin/adhoc-scripts/updatePileupObjects.py [hunk 6]
```diff
@@ -147,6 +158,65 @@ def writePileupDocs(mspileupUrl, puDocs, handler, logger):
...
+    puDocs = getPileupDocs(url, handler, logger)
+    logger.info("Found %d documents in MSPileup", len(puDocs))
+
+    # update the documents with the override content
+    for doc in puDocs:
+        doc.update(puOverride)
+        logger.info("New pileup document is: %s", doc)
+
+    if dryrun:
+        logger.info("documents are not written due to --dry-run option\n")
```
Comment (amaltaro) [inline bin/adhoc-scripts/updatePileupObjects.py:213]: If it's dry-run, should we print the documents such that they can be inspected before actually writing them back into the database?

File: bin/adhoc-scripts/updatePileupObjects.py [hunk 7]
```diff
@@ -147,6 +158,65 @@ def writePileupDocs(mspileupUrl, puDocs, handler, logger):
...
+    """
+    helper function to update pileup records
+    :param handler: HTTP handler to use
+    :param url: url to use (string)
+    :param fin: input file name to read from (string)
+    :param logger: logger object to use
+    :param dryrun: option to run dry-run mode (boolean)
+    :return: nothing
+    """
+    with open(fin, 'r', encoding='utf-8') as istream:
```
Comment (amaltaro) [inline bin/adhoc-scripts/updatePileupObjects.py:200]: It looks like that with the current changes, we should stop making `fin` required in this script, no? In other words, I see 2 scenarios:
a) we provide a json file as input and write (update) them in the database
b) we do not provide a json file, fetch documents via MSPileup, update them in memory and persist them back in the database.

File: bin/adhoc-scripts/adjustMongoDocs.py [hunk 8]
```diff
@@ -40,34 +47,80 @@ def __init__(self):
...
+    """
+    Helper function to add and update in place transition record for existing MSPileup records
+    :param dburi: database URI (string)
+    :param dbname: database name (string)
+    :param dbcoll: database collection name (string)
+    :param userDN: string representing user DN
+    :param verbose: verbose flag (boolean)
+    :return: nothing
+    """
+    userDN = '/DC=ch/DC=cern/OU=Organic Units/OU=Users/CN=cmsunified/CN=658085/CN=Robot: CMS Unified account'
```
Comment (amaltaro) [inline bin/adhoc-scripts/adjustMongoDocs.py:68]: I think we need to delete this line and rely on the `defaultDN`, defined in the `__init__` method.

**Review by amaltaro — CHANGES_REQUESTED** [review]
Valentin, in addition to the comments along the code, please also update the PR description (as now you are providing changes to 2 scripts).

File: bin/adhoc-scripts/updatePileupObjects.py (hunk 6 above)
Comment (vkuznet) [reply bin/adhoc-scripts/updatePileupObjects.py]: ok, I added printout of docs.

File: bin/adhoc-scripts/updatePileupObjects.py (hunk 7 above)
Comment (vkuznet) [reply bin/adhoc-scripts/updatePileupObjects.py]: I think both options are valid use-case and I adjusted the code to support them both.

File: bin/adhoc-scripts/adjustMongoDocs.py (hunk 8 above)
Comment (vkuznet) [reply bin/adhoc-scripts/adjustMongoDocs.py]: ok, done.

**Comment by vkuznet** [comment]
Alan, I updated the code and description, please have a final look.

File: bin/adhoc-scripts/updatePileupObjects.py [hunk 9]
```diff
@@ -147,6 +158,68 @@ def writePileupDocs(mspileupUrl, puDocs, handler, logger):
...
+        puDocs = getPileupDocs(url, handler, logger)
+        logger.info("Found %d documents in MSPileup", len(puDocs))
+
+    # update the documents with the override content
+    for doc in puDocs:
+        doc.update(puOverride)
+        logger.info("New pileup document is: %s", doc)
+
+    if dryrun:
+        logger.info("Updated documents")
```
Comment (amaltaro) [inline bin/adhoc-scripts/updatePileupObjects.py:214]: Please fix this message not to mislead with real update action (this is a dry run!)

File: bin/adhoc-scripts/updatePileupObjects.py [hunk 10]
```diff
@@ -147,6 +158,68 @@ def writePileupDocs(mspileupUrl, puDocs, handler, logger):
...
+    """
+    helper function to update pileup records
+    :param handler: HTTP handler to use
+    :param url: url to use (string)
+    :param fin: input file name to read from (string)
+    :param logger: logger object to use
+    :param dryrun: option to run dry-run mode (boolean)
+    :return: nothing
+    """
+    if os.path.exists(fin):
```
Comment (amaltaro) [inline bin/adhoc-scripts/updatePileupObjects.py:200]: I think the option parser needs to be updated and input file to become an optional command line argument.

**Review by amaltaro — CHANGES_REQUESTED** [review]

File: bin/adhoc-scripts/updatePileupObjects.py (hunk 9 above)
Comment (vkuznet) [reply bin/adhoc-scripts/updatePileupObjects.py]: I made it more clear.

File: bin/adhoc-scripts/updatePileupObjects.py (hunk 10 above)
Comment (vkuznet) [reply bin/adhoc-scripts/updatePileupObjects.py:200]: fixed.

Comment (amaltaro) [reply bin/adhoc-scripts/updatePileupObjects.py:200]: And yet you are checking if the file exists, regardless whether an input was passed in or not. This is definitely wrong.

**Review by amaltaro — CHANGES_REQUESTED** [review]

File: bin/adhoc-scripts/updatePileupObjects.py (hunk 10 above)
Comment (vkuznet) [reply bin/adhoc-scripts/updatePileupObjects.py:200]: Why checking for file presence is wrong? What if I had a typo in provided argument. How about changing this to the following construct:
```
if fin:
    if os.path.exists(fin):
         # do action
    else:
         logger.error("Provided file %s does not exist", fin)
         os.exit(1)
else:
    # read docs from MongoDB
```

Comment (amaltaro) [reply bin/adhoc-scripts/updatePileupObjects.py:200]: Precisely my point, we should only check for file existence if a file/path name is provided.

Comment (vkuznet) [reply bin/adhoc-scripts/updatePileupObjects.py]: ok, cleared.

**Review by amaltaro — APPROVED** [review]
It looks good to me. Before merging it, can you please update the documents in testbed MSPileup? Upon testing and verifying that, we can proceed with merging this and also updating the production documents.

**Comment by vkuznet** [comment]
Alan, upon testing the script against testbed I found new flaws in its logic. The main issue was the following:
- we get all docs from MSPileup service
- we update records without transition record
- **but we used all docs again to write back to MSPileup**

The last step did a wrong thing since when we provided a document with already existing transition record the MSPileup code itself run the logic to use last transition record to create custom pileup name, etc. This is why I saw documents with such kind of structures:
```
"transition": [
    {
      "containerFraction": 1.0,
      "customDID": "/hasan-test10-Neutrino_E-10_gun/Run3Summer21PrePremix-120X_mcRun3_2021_realistic_v6-v2/PREMIX",
      "updateTime": 1709576366,
      "DN": "/DC=ch/DC=cern/OU=Organic Units/OU=Users/CN=msunmer/CN=852819/CN=Robot: WmCore Service Account"
    },
    {
      "containerFraction": 1.0,
      "customDID": "/hasan-test10-Neutrino_E-10_gun/Run3Summer21PrePremix-120X_mcRun3_2021_realistic_v6-v2/PREMIX-V1",
      "updateTime": 1709579969,
      "DN": "/DC=ch/DC=cern/OU=Organic Units/OU=Users/CN=valya/CN=443502/CN=Valentin Y Kuznetsov"
    },
    {
      "containerFraction": 1.0,
      "customDID": "/hasan-test10-Neutrino_E-10_gun/Run3Summer21PrePremix-120X_mcRun3_2021_realistic_v6-v2/PREMIX-V2",
      "updateTime": 1709581549,
      "DN": ""
    },
    {
      "containerFraction": 1.0,
      "customDID": "/hasan-test10-Neutrino_E-10_gun/Run3Summer21PrePremix-120X_mcRun3_2021_realistic_v6-v2/PREMIX-V3",
      "updateTime": 1709585151,
      "DN": ""
    }
  ]
```
In other words, we should never pass **all** record for the update and only pass those which have not had any transition record. I fixed this and other issue.

The other issue I found is when I tried to use `--fin` parameter which contain a change we want to apply to the document. I tried to reset transition records to empty list but since on testbed we already have code which checks transition my attempt fail.

I left my last commits without squashing that you can easily review them. Please do that:
- https://github.com/dmwm/WMCore/pull/11915/commits/eb15f24ca9a4383c40e398c3fbf78f445614a437 fixes condition in adjust script (to be aligned with update script code)
- https://github.com/dmwm/WMCore/pull/11915/commits/4cf9f1e9e79d6454923e9796a726a1bc0c073d26 fixes logic of of update (both issues I listed above)
- https://github.com/dmwm/WMCore/pull/11915/commits/f34e44a5288f8451530218d7deccbb55052932ce clarifies meaning of fin option to avoid confusion.

**Comment by amaltaro** [comment]
@vkuznet Valentin, if I understood this statement right:
""In other words, we should never pass all record for the update and only pass those which have not had any transition record. I fixed this and other issue.""

and the example of transition records keeping the same container fraction of `1.0`, I would say this is a misbehavior on the server side (MSPileup) which we need to fix. We don't want the service to create custom container UNLESS there was a change of container fraction.

**Comment by vkuznet** [comment]
Yes, it seems it is bug in server side, but I can't be sure since I don't know which code is deployed on testbed. We had so many issues/PRs that it is frankly possible we run not the latest version which may not have this issue. Said that, I agree to file new issue to check this.

**Comment by amaltaro** [comment]
Valentin, you can check the service version through this REST API:
https://cmsweb-testbed.cern.ch/ms-pileup/data/info

and see what changes went in through the CHANGES file:
https://github.com/dmwm/WMCore/blob/master/CHANGES

Said that, testbed is running `2.3.1rc4`, which is exactly the same as `2.3.1` deployed in production.

In summary, the services version we have in testbed (and production) are supposed to have 100% of this feature.

**Comment by vkuznet** [comment]
Upon further investigation, now I see that in fact, the reported here transition records https://github.com/dmwm/WMCore/pull/11915#issuecomment-1978754084 are correctly created since the entire MSPileup record contains container fraction 0.5 while I send a document with transition record 1.0. That's why we start see growth of transition records. But other existing records where container fraction was 1.0 have not had this issue. The remaining mystery for me right now how empty DN appears.

**Comment by vkuznet** [comment]
Ok, now I understand where empty DN comes from too. It comes from `MSPileupTask.py` which calls `updatePileup` method of `MSPileupData.py` module without providing userDN which is optional parameter of latter methon with empty string default value. Therefore, the bug we have is in `MSPileupTask.py` module where we should properly provide userDN which will be recorded in transition record. How to do it is yet unclear to me.

File: bin/adhoc-scripts/updatePileupObjects.py [hunk 11]
```diff
@@ -19,20 +22,28 @@
...
+# WMCore services
+from Utils.Timers import gmtimeSeconds
+
 
 class OptionParser():
     """Class to parse the command line arguments"""
 
     def __init__(self):
         "User based option parser"
+        defaultDN = '/DC=ch/DC=cern/OU=Organic Units/OU=Users/CN=msunmer/CN=852819/CN=Robot: WmCore Service Account'
```
Comment (amaltaro) [inline bin/adhoc-scripts/updatePileupObjects.py:34]: I actually think we should not persist any DNs in github. So I would suggest to fix the relevant records in MongoDB (testbed and production), then we update this pull request before merging it.

File: bin/adhoc-scripts/adjustMongoDocs.py [hunk 12]
```diff
@@ -25,12 +28,16 @@
...
+# WMCore services
+from Utils.Timers import gmtimeSeconds
+
 
 class OptionParser(object):
     "Helper class to handle script options"
 
     def __init__(self):
         "User based option parser"
+        defaultDN = '/DC=ch/DC=cern/OU=Organic Units/OU=Users/CN=msunmer/CN=852819/CN=Robot: WmCore Service Account'
```
Comment (amaltaro) [inline bin/adhoc-scripts/adjustMongoDocs.py:40]: Same comment here.

Comment (vkuznet) [reply bin/adhoc-scripts/adjustMongoDocs.py]: removed

File: bin/adhoc-scripts/updatePileupObjects.py (hunk 11 above)
Comment (vkuznet) [reply bin/adhoc-scripts/updatePileupObjects.py]: removed

**Comment by vkuznet** [comment]
Alan, I don't know what else needs to be done here, but I removed default DN since it can be provided as external parameter to the script. Therefore, I request final review and if you're satisfied I can squash changes for the merge.

**Review by amaltaro — APPROVED** [review]
Everything looks good to me, but I would like to first update the transition records with this script, before we actually merge it.

It looks like testbed has the expected transition record now, right? Do I understand it right that first we need to upgrade MSPileup, before we can actually update the production records?

**Comment by vkuznet** [comment]
Cmsweb testbed have already transition records. But the MSPileupData code checks if transition record is in place, see https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/MicroService/MSPileup/MSPileupData.py#L259, before allowing the update. Therefore, I think we first need transition record to be in place in production MongoDB then we can setup new code.

**Comment by amaltaro** [comment]
> Therefore, I think we first need transition record to be in place in production MongoDB then we can setup new code.

Okay. Can you then update the production records with the script provided in this PR? Once we verify that, we can merge it.
But we need to keep in mind that new pileup documents might be created until we actually upgrade services in production, so we will have to run the same procedure once again (please keep/share notes on how/where exactly to update mongodb records).

**Comment by vkuznet** [comment]
per our discussion on MM, I'll wait until further notice to update production MSPileup records.

**Comment by amaltaro** [comment]
Before merging it, please squash the commits. @vkuznet

**Comment by amaltaro** [comment]
@vkuznet just a reminder about squashing commits.

**Comment by vkuznet** [comment]
Alan, I made a mistake locally by squashing other parts within this PR, so you have two options:
- if you want squashed commits I can recreate PR with another branch and PR
- or, take it as is w/o squashing commits.

Sorry about mistake.

**Comment by amaltaro** [comment]
No problem. Please create a new PR and write that it superseeds this one.

**Comment by vkuznet** [comment]
Due to mistake during squash local procedure (I squashed an additional changes not relevant to this PR) I created a new PR#11947 which supersede this PR.

**Comment by amaltaro** [comment]
Superseeded by https://github.com/dmwm/WMCore/pull/11947
I am closing this one out, but please reopen it if I missed anything, Valentin.

**Pull request closed without merging**